compiler/rustc_span/src/lib.rs

```diff
@@ -192,13 +223,7 @@ impl std::fmt::Display for FileName {
     fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         use FileName::*;
         match *self {
-            Real(RealFileName::Named(ref path)) => write!(fmt, "{}", path.display()),
-            // FIXME: might be nice to display both components of Devirtualized.
-            // But for now (to backport fix for issue #70924), best to not
-            // perturb diagnostics so its obvious test suite still works.
-            Real(RealFileName::Devirtualized { ref local_path, virtual_name: _ }) => {
-                write!(fmt, "{}", local_path.display())
-            }
+            Real(ref name) => write!(fmt, "{}", name.most_stable_name().display()),
```

Comment:
I don't think we should change the semantics of this implementation. Do you have an idea how much impact it would have to remove the `Display` impl for `Filename` altogether? I'd like to try forcing `Filename` users to explicitly make a decision which version of the path they want. For example by adding the `most_stable_name` method to `FileName` too and a `local_name_if_available()` method. The values returned by each of these would then implement `Display`.

Reply:
So I chased down every usage of `FileName::fmt` and `to_string`. The commits after https://github.com/rust-lang/rust/pull/83813/commits/3047fc7ae7b8985aad85bbdb9d04a8a39d3fc287 are the work.

`FileName::to_string_virtual` is used when it isn't a `FileName::Real(_)` and we don't have to worry about paths. This is helpful sometimes when the callsite had already matched on `Real(_)` and we just need an irrefutable "`otherwise`" arm. Panics if called on `Real(_)`. Might be able to lift it to compile time checks once https://github.com/rust-lang/rfcs/pull/2593 is implemented & stabalised but that's a long way off.

`FileName::to_string_local` is "local name if available". Only to be used when it won't end up in build output.

`FileName::to_string_embeddable` is used when the path will be embedded. This is basically a wrapper around `most_stable_name`.

I swapped `impl Display` methods for the above three functions in steps. `to_string_virtual` is the easiest, immediately obvious from the callsite: https://github.com/rust-lang/rust/pull/83813/commits/8c9214403d5d979eb450e93bdb6753ea77e01e2d (some changes are in the wrong commit due to me trying to rewrite history)

And then some places definitely needs `to_string_embeddable`: https://github.com/rust-lang/rust/pull/83813/commits/0d53c343826a52d3874f0e30fe7fcfaffe4212ef. Mostly codegen backends.

Some places are fine with `to_string_local`: https://github.com/rust-lang/rust/pull/83813/commits/2507bce5b8d7cec31491c05aa2948d2a061ce771, mostly error outputs and rustdoc (which should never get a `Remapped` variant I think)

The rest I'm not sure about. In this case `to_string_embeddable` is always the safer option. Wrong application might make some diagnostics slightly confusing but that's better than leaking information.

`impl Display` is now removed for `FileName` so every place makes an explicit choice.

Reply:
Yes, the amount of change required for removing the `Display` impl seems quite acceptable. It it will certainly make it easier to catch errors now and in the future. I'm very pleased with that.

I think we can improve the setup to be a bit more ergonomic and efficient. I see three specific issues:
- We now have a number of terms now ("local", "embeddable", "remapped", and "most_stable") that somewhat overlap. We should reduce the redundancy here in order to avoid confusion. 
- In the current state of the PR we unconditionally allocate memory for a `String` but in most cases we can actually avoid that. 
- From a user's perspective it does not matter very much if a `FileName` is virtual or not (at least not when printing/embedding them). So I think it's better to remove the `to_string_virtual()` method.

I propose that we create a thin wrapper type for `FileName` for printing and embedding purposes. This type can then implement `Display` (for use in `format!()` etc) and provide a `to_string_lossy(&self) -> Cow<str>` method (for use with `Symbol::intern()`). Something like:

```rust
struct FileNameDisplay {
    inner: &FileName,
    prefer_local: bool,
}

impl Display for FileNameDisplay {
     // ...
}

impl FileNameDisplay {
    fn to_string_lossy(&self) -> Cow<'_, str> {
        match self.inner {
            FileName::Real(ref inner) => inner.to_string_lossy(self.prefer_local),
            other => Cow::from(format!("{}", self)),
        }
    }
}

impl FileName {
     fn prefer_remapped() -> FileNameDisplay { ... }
     fn prefer_local() -> FileNameDisplay { ... }
}
``` 

`RealFileName` would then probably also get a `to_string_lossy()` method with a `prefer_local: bool` parameter (as used above). In turn the `most_stable_name()` method (and the "most stable" terminology in general) could be removed. Then we could call `Symbol::intern()` without allocating memory for most real file names, e.g.:

```rust
Symbol::intern(&loc.file.name.prefer_remapped().to_string_lossy())
```

The `to_string_virtual` method would go away. One would just use `prefer_local()` or `prefer_remapped()` as appropriate (and get the same result in both cases). Alternatively, the method could be renamed to something like `to_string_assert_is_virtual()`.

Reply:
I renamed `most_stable_name()` to `remapped_name_if_available()` because there are places needing an embeddable `Path`. If we were to remove it we'd have to convert to `String` and back again.

Reply:
Those changes look good to me. Now we just need to figure out a clean handling for `StableSourceFileId` ...